### ISSUE TITLE
[css-2024] Add features to Safe to Release pre-CR Exceptions

### ISSUE THREAD

**Issue by SebastianZ** [w3c/csswg-drafts#10687]
Here's a (very hand-wavy) list of features I believe could be added to the "Safe to Release pre-CR Exceptions" section of CSS 2024:

* All from CSS Animations 2
* `font-tech()` and `font-format()` from CSS Conditional 5
* `linear()` from CSS Easing 2 (requires to publish a FPWD)
* `env()` from CSS Environment Variables 1 (still some open issues but interoperably implemented, requires to publish a FPWD)
* All from CSS Highlight API
* All from CSS Images 4
* `text-box-trim` and `text-box-edge` from CSS Inline Layout 3
* All from CSS Overflow 3
* All from CSS Paged Media 3
* Typographic pseudo-elements, `::before`, `::after` from CSS Pseudo-Elements 4 (basically all from CSS 2 with double-colon syntax plus `::prefix` and `::postfix`)
* `:is()`, `:where()`, `:has(), and selector lists in `:not()` from Selectors 4 ([Selectors 4 is also suggested to be added to "Rough Interoperability"](https://github.com/w3c/csswg-drafts/issues/9793).)
* `text-decoration-thickness`, `text-underline-offset` properties and `from-font` value for `text-underline-position` from CSS Text Decorations 4
* `transition-behavior` property, `@starting-style` rule and CSSStartingStyleRule interface of CSS Transitions 2
* `accent-color` property,`auto` value for `outline-color` property, and `<url-set>` value for the `cursor` property from CSS UI 4
* Several of the interfaces defined in CSSOM 1 (which I believe should be cut now regarding its feature set and a CSSOM 2 spec. started for new features)

For this list I went through all the specs and checked what seems fairly stable to me based on issues within the spec. and on GitHub and/or is already widely implemented in browsers. For some I checked the WPT coverage, though not all. I may go through them and provide some more detailed info. And there are surely some [Interop 2023](https://wpt.fyi/interop-2023) and [2024](https://wpt.fyi/interop-2024) features I didn't list here yet, which might be added.

Everyone is encouraged to edit this list or suggest additions, changes, or removals! And let me know if it makes sense to split it into separate issues to discuss them individually!

Sebastian

**Comment by svgeesus** [comment]
> * `linear()` from CSS Easing 2 (requires to publish a FPWD)

Yes we absolutely need a FPWD :) also it would be good if the spec had WPT annoutations to show [how well it is tested](https://wpt.fyi/results/css/css-easing?label=master&label=experimental&aligned&q=linear)

**Comment by svgeesus** [comment]
In general, if the proposal for pre-CR exceptions is "all from spec foo" then it kind of implies that foo should move to CR.

**Comment by SebastianZ** [comment]
> In general, if the proposal for pre-CR exceptions is "all from spec foo" then it kind of implies that foo should move to CR.

Agree, though some of them have a lot of open issues (in-spec and/or on GitHub), but all of their features are widely implemented. Not sure how to handle those. I'll try to dig up some numbers, so we can get a better picture of their actual state, though that'll take a bit.

Sebastian

**Comment by svgeesus** [comment]
> Yes we absolutely need a FPWD :)

To keep this issue up to date, the [FPWD of CSS Easing 2](https://www.w3.org/TR/2024/WD-css-easing-2-20240829/) was published 29 August 2024 and I did add the WPT annotations.

**Comment by svgeesus** [comment]
> * `env()` from CSS Environment Variables 1 (still some open issues but interoperably implemented, requires to publish a FPWD)

We still don't have a FPWD yet.

**Comment by SebastianZ** [comment]
I have now updated the list of features and added additional information and stats to them. The info on some of them is still incomplete and I'll continue updating it.
Though I believe we're now in a state to discuss this. So adding it to the agenda.

Sebastian

**Comment by fantasai** [comment]
- text-box* needs some edits, it's close, but it's not quite at "ready to ship". We should track it independently.
- anything in CSS2 doesn't need to be listed, it's already in REC

**Comment by astearns** [comment]
There is a question on whether we should include `interactivity` in this list, as the html spec will be referencing it 

https://drafts.csswg.org/css-ui-4/#inertness

**Comment by svgeesus** [comment]
> > * `linear()` from CSS Easing 2 (requires to publish a FPWD)
> 
> Yes we absolutely need a FPWD :) also it would be good if the spec had WPT annoutations to show [how well it is tested](https://wpt.fyi/results/css/css-easing?label=master&label=experimental&aligned&q=linear)

The FPWD of Easing 2 ws [published 29 August 2024](https://www.w3.org/TR/2024/WD-css-easing-2-20240829/) and includes WPT annotations

**Comment by SebastianZ** [comment]
> The FPWD of Easing 2 ws [published 29 August 2024](https://www.w3.org/TR/2024/WD-css-easing-2-20240829/) and includes WPT annotations

You already [mentioned that earlier](https://github.com/w3c/csswg-drafts/issues/10687#issuecomment-2458174602).😃

Sebastian

**Comment by SebastianZ** [comment]
> > * `env()` from CSS Environment Variables 1 (still some open issues but interoperably implemented, requires to publish a FPWD)
> 
> We still don't have a FPWD yet.

I've created #11899, which is blocking #6729.

Sebastian

**Comment by SebastianZ** [comment]
I've updated "font-tech() and font-format() from CSS Conditional 5", "linear() from CSS Easing 2", "env(safe-area-inset-*) from CSS Environment Variables 1", and "All from CSS Highlight API 1".
More to come.

Sebastian

**Comment by SebastianZ** [comment]
Updated "conic-gradient() and <color-interpolation-method> for gradient functions from CSS Images 4", "text-box-trim and text-box-edge from CSS Inline Layout 3", and "All from CSS Overflow 3".

Sebastian

**Comment by SebastianZ** [comment]
Updated "All from CSS Overflow 3", "All from CSS Paged Media 3" (though now wondering a bit whether really _all_ should be added), "Typographic pseudo-elements, ::before, ::after from CSS Pseudo-Elements 4", and "text-decoration-thickness, text-underline-offset properties and from-font value for text-underline-position from CSS Text Decorations 4".

Sebastian

**Comment by SebastianZ** [comment]
Updated the rest and included additional WPT and issues info.

Sebastian

**Comment by SebastianZ** [comment]
Updated data on CSS Overflow 3 and reorganized features from stable to less stable.

Sebastian

**Comment by svgeesus** [comment]
For CSS Images 4:
  
> * [WPT](https://wpt.fyi/results/css/css-images?label=master&label=stable&aligned&q=gradient): passing tests: Chrome: 2518, Edge: 2518, Firefox: 2897, Safari: 2857; [interop](https://wpt.fyi/results/css/css-images?label=master&label=stable&aligned&q=gradient+and+all(status%3Apass)): 2511 / 2907 (86.4%)

 Currently Chrome 2953 | Firefox 3335 | Safari 3249 | Edge 2963

Now [100% interop](https://wpt.fyi/results/css/css-images?label=stable&label=master&aligned&q=gradient%20and%20all%28status%3Apass%29) (2559 / 2559) on those gradient tests.

Also

 - https://github.com/w3c/csswg-drafts/issues/12756

**Comment by SebastianZ** [comment]
Updated info and stats for typographic pseudo-elements, `font-tech()` and `font-format()`, `:is()`, `:where()`, `:has()`, and selector lists in `:not()`, and `conic-gradient()` and `<color-interpolation-method>`.


@svgeesus:

> For CSS Images 4:
> 
> > * [WPT](https://wpt.fyi/results/css/css-images?label=master&label=stable&aligned&q=gradient): passing tests: Chrome: 2518, Edge: 2518, Firefox: 2897, Safari: 2857; [interop](https://wpt.fyi/results/css/css-images?label=master&label=stable&aligned&q=gradient+and+all(status%3Apass)): 2511 / 2907 (86.4%)
> 
> Currently Chrome 2953 | Firefox 3335 | Safari 3249 | Edge 2963
> 
> Now [100% interop](https://wpt.fyi/results/css/css-images?label=stable&label=master&aligned&q=gradient%20and%20all%28status%3Apass%29) (2559 / 2559) on those gradient tests.

The interop statistics are always 100%, as that's only showing the tests that pass in all browsers. The actual interop value is the number of tests that pass in all browsers devided by the total amount of tests, which is 2435 / 2948 (82.6%) in case of the tests related to conic gradients and the interpolation methods. Though note that the actual value might be a little different, as it is pretty hard to restrict the tests to a specific feature in the WPT search.

Sebastian

**Comment by svgeesus** [comment]
> The interop statistics are always 100%, as that's only showing the tests that pass in all browsers. The actual interop value is the number of tests that pass in all browsers devided by the total amount of tests

D'oh! Sorry.

**Comment by SebastianZ** [comment]
Updated info and stats for `linear()`, `transition-behavior` property, `@starting-style` rule and `CSSStartingStyleRule` interface, and `env(safe-area-inset-*)`.

Sebastian

**Comment by SebastianZ** [comment]
Updated info and stats for the `accent-color` property, `auto` value for `outline-color` property, and `<url-set>` value for `cursor`, CSS Highlight API 1, `text-box-trim` and `text-box-edge`, CSS Paged Media 3, and CSSOM 1.

Sebastian

**Comment by svgeesus** [comment]
> [`conic-gradient()`](https://drafts.csswg.org/css-images-4/#conic-gradients) and [`<color-interpolation-method>` for gradient functions](https://drafts.csswg.org/css-images-4/#:~:text=Added+%3Ccolor-interpolation-method%3E+to+all+gradient+functions) from [CSS Images 4](https://drafts.csswg.org/css-images-4/)

While preparing the document for publication I ran a diff between the current one and the one published last week.

I noticed that we already had, [from CSS 2024](https://www.w3.org/TR/css-2024/#CR-exceptions):

> The [conic-gradient()](https://www.w3.org/TR/css-images-4/#funcdef-conic-gradient) gradient notation. See [decision](https://github.com/w3c/csswg-drafts/issues/2383#issuecomment-371340088).

and

> The [&lt;color-interpolation-method>](https://www.w3.org/TR/css-color-5/#color-interpolation-method), defined in [[CSS-COLOR-4]](https://www.w3.org/TR/css-2024/#biblio-css-color-4) and used for interpolation of linear, radial and conic gradients. [[CSS-IMAGES-4]](https://www.w3.org/TR/css-2024/#biblio-css-images-4)

so I will remove the recent addition of these, which is a duplicate.
